providers/default_rhel.rb

```diff
@@ -0,0 +1,136 @@
+#
+# Cookbook Name:: kernel-modules
+# Author:: Jeremy MAURO <j.mauro@criteo.com>
+#
+# Copyright 2016, Criteo.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+provides :kernel_module, platform_family: 'rhel'
+
+MODPROBE_BIN = '/sbin/modprobe'.freeze
+
+def whyrun_supported?
+  true
+end
+
+use_inline_resources
+
+action :configure do
+  # Requirements
+  package node['kernel_modules']['packages']
+
+  # Module init loading section
+  template modload_file do
+    source "#{node['platform_family']}-#{node['platform_version'].to_i}/init-load.erb"
+    cookbook 'kernel-modules'
+    mode '0755'
+    variables(
+      name: new_resource.name,
+      bin: MODPROBE_BIN,
+    )
+    action new_resource.onboot ? :create : :delete
+  end
+
+  # Module loading options
+  template modprobe_file do
+    source 'modprobe.conf.erb'
+    cookbook 'kernel-modules'
+    mode '0644'
+    variables(
+      name: new_resource.name,
+      alias: new_resource.alias,
+      options: new_resource.options,
+      install: new_resource.install,
+      remove: new_resource.remove,
+      blacklist: new_resource.blacklist,
+    )
+    action :create
+  end
+end
+
+action :load do
+  # When loading a module it means the module must be configured
+  run_action :configure
+
+  wanted_action = blacklisted?(new_resource.install, new_resource.blacklist) ? :run : :nothing
+  execute "Unloading module #{new_resource.name}" do
+    command reload_cmd
+    only_if "lsmod | grep -q #{new_resource.name}"
+    only_if { new_resource.reload || new_resource.force_reload } # "force_reload" supersede "reload" property
+
+    subscribes :run, "template[#{modprobe_file}]" # 'Unload' only_if template has been updated
+    action wanted_action # If module blacklisted unload no matter what
+  end
+
+  execute "(Re)Loading module #{new_resource.name}" do
+    command "#{MODPROBE_BIN} -v #{new_resource.name}"
+    not_if "lsmod | grep -q #{new_resource.name}"
+    not_if { wanted_action.to_s == 'run' } # No need to load if blacklisted
+  end
+end
+
+action :unload do
+  # Requirements
+  package node['kernel_modules']['packages']
+
+  execute "Unloading module #{new_resource.name}" do
+    command reload_cmd
+    only_if "lsmod | grep -q #{new_resource.name}"
+    only_if { new_resource.reload || new_resource.force_reload } # "force_reload" supersede "reload" property
```

Comment:
No we want to unload even if reload is false. Reload property is only to trigger reload if config changes